python/cudf/cudf/core/series.py

```diff
@@ -815,6 +817,100 @@ def __array_function__(self, func, types, args, kwargs):
         else:
             return NotImplemented
 
+    def map(self, arg, na_action=None) -> "Series":
+        """
+        Map values of Series according to input correspondence.
+
+        Used for substituting each value in a Series with another value,
+        that may be derived from a function, a ``dict`` or
+        a :class:`Series`.
+
+        Parameters
+        ----------
+        arg : function, collections.abc.Mapping subclass or Series
+            Mapping correspondence.
+        na_action : {None, 'ignore'}, default None
+            If 'ignore', propagate NaN values, without passing them to the
+            mapping correspondence.
+
+        Returns
+        -------
+        Series
+            Same index as caller.
+
+        Examples
+        --------
+        >>> s = cudf.Series(['cat', 'dog', np.nan, 'rabbit'])
+        >>> s
+        0      cat
+        1      dog
+        2      NaN
+        3   rabbit
+        dtype: object
+
+        ``map`` accepts a ``dict`` or a ``Series``. Values that are not found
+        in the ``dict`` are converted to ``NaN``, unless the dict has a default
+        value (e.g. ``defaultdict``):
+
+        >>> s.map({'cat': 'kitten', 'dog': 'puppy'})
+        0   kitten
+        1    puppy
+        2      NaN
+        3      NaN
+        dtype: object
+
+        It also accepts numeric functions:
+        >>> s = cudf.Series([1, 2, 3, 4, np.nan])
+        >>> s.map(lambda x: x ** 2)
+        0       1
+        1       4
+        2       9
+        3       16
+        4       NaN
+        dtype: int64
+
+        Please note map does not currently support user defined functions
+        with dtypes `str` and`category`.
+        """
+        if isinstance(arg, dict):
+            lhs = cudf.DataFrame({"x": self, "orig_order": arange(len(self))})
+            rhs = cudf.DataFrame(
+                {
+                    "x": arg.keys(),
+                    "s": arg.values(),
+                    "bool": full(len(arg), True, dtype=self.dtype),
+                }
+            )
+            res = lhs.merge(rhs, on="x", how="left").sort_values(
+                by="orig_order"
+            )
+            res = res.drop_duplicates(subset="orig_order", ignore_index=True)
+            result = res["s"]
+            result.name = self.name
+        elif isinstance(arg, cudf.Series):
+            lhs = cudf.DataFrame({"x": self, "orig_order": arange(len(self))})
+            rhs = cudf.DataFrame(
+                {
+                    "x": arg.keys(),
+                    "s": arg,
+                    "bool": full(len(arg), True, dtype=self.dtype),
+                }
+            )
+            res = lhs.merge(rhs, on="x", how="left").sort_values(
+                by="orig_order"
+            )
+            res = res.drop_duplicates(subset="orig_order", ignore_index=True)
```

Comment:
This would be to handle the situation where `arg.keys()` (the Series index) isn't unique so it generates multiple matches per row of the `lhs` dataframe.

Pandas actually throws if `arg.keys()` doesn't have unique values:
```python
import pandas as pd

a = pd.Series([1, 1, 2, 3, 4])
b = pd.Series(['a', 'b', 'c'], index=[1, 1, 2])

a.map(b)
```
```python-traceback
----> 1 a.map(b)

~/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/series.py in map(self, arg, na_action)
   3968         dtype: object
   3969         """
-> 3970         new_values = super()._map_values(arg, na_action=na_action)
   3971         return self._constructor(new_values, index=self.index).__finalize__(
   3972             self, method="map"

~/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/base.py in _map_values(self, mapper, na_action)
   1129             values = self._values
   1130
-> 1131             indexer = mapper.index.get_indexer(values)
   1132             new_values = algorithms.take_1d(mapper._values, indexer)
   1133

~/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/indexes/base.py in get_indexer(self, target, method, limit, tolerance)
   2979         if not self.is_unique:
   2980             raise InvalidIndexError(
-> 2981                 "Reindexing only valid with uniquely valued Index objects"
   2982             )
   2983

InvalidIndexError: Reindexing only valid with uniquely valued Index objects
```

Reply:
This is helpful to know! Should I be checking for this and throwing an error as well?

Reply:
FWIW when I run this at the moment, we get a length mismatch error

Reply:
Yes, we should check this and throw an error as well. The reason you're getting a length mismatch error is because you'll end up generating multiple matches in the case of `1`. So while your input series is length 5, your output series would be length 7.

Reply:
That's also very helpful to know! I'll add a check for this :D

Reply:
Added a check for this :D